File: cs480/astar/java/AStar.java
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

public abstract class AStar {

    Object goalNode;
    AStarFringe fringe;
    Map<String, Integer> minNodeScore;
    Map<String, Object> cameFrom;
    int nodesVisited;

    public AStar(Object rootNode, Object goalNode) {
        this.goalNode = goalNode;

        this.fringe = new AStarFringe();
        this.fringe.addNode( rootNode, this.hScore(rootNode) );

        this.minNodeScore = new HashMap<String, Integer>();
        this.minNodeScore.put( rootNode.toString(), 0 );

        this.cameFrom = new HashMap<String, Object>();

        this.nodesVisited = 0;
    }

    public abstract List<NodeAndCost> children(Object node);
    public abstract Integer hScore(Object node);

    public List<Object> search() {
        while ( !this.fringe.isEmpty() ) {
            Object currentNode = this.fringe.popNode();
            this.nodesVisited++;

            if (this.isGoal(currentNode)) {
                return this.pathToRoot(currentNode);
            }
            else {
                this.addChildrenToFringe(currentNode);
            }
        }
        return null;
    }

    public void printPath(List<Object> path) {
        for (Object node : path) {
            System.out.println(node.toString());
        }
    }

    private List<Object> pathToRoot(Object node) {
        List<Object> path = new ArrayList<Object>();
        Object curNode = node;
        do {
            path.add(curNode);
            curNode = this.cameFrom.get(curNode.toString());
        } while ( curNode != null);

        Collections.reverse(path);

        return path;
    }

    private boolean isGoal(Object node) {
        return node.toString().equals(this.goalNode.toString());
    }

    private void addChildrenToFringe(Object node) {
        String nodeKey = node.toString();
        Integer costToNode = this.minNodeScore.get(nodeKey);

        for (NodeAndCost childAndCost : this.children(node)) {
            Integer costToChild = costToNode + childAndCost.cost;
            String childKey = childAndCost.node.toString();

            Integer curChildCost = this.minNodeScore.get(childKey);
            if (curChildCost == null || costToChild < curChildCost) {
                this.fringe.addNode(childAndCost.node, costToChild + this.hScore(childAndCost.node));
                this.minNodeScore.put(childKey, costToChild);
                this.cameFrom.put(childKey, node);
            }
        }
    }

    public class NodeAndCost {
        Object node;
        Integer cost;

        public NodeAndCost(Object node, Integer cost) {
            this.node = node;
            this.cost = cost;
        }
    }
}

File: cs480/astar/project1/astar.py
from priority_queue import PriorityQueue

class AStar(object):
    """
    Given a starting node and a goal node, finds the shortest cost path
    from the starting node to the goal node.
    """

    def __init__(self, root_node, goal_node):
        """
        Initializes the AStar object and sets function that determines the
        order in which the nodes are evaluated.

        Precondition: nodes can be any type of object, but must be able to
        be compared by str(node1) == str(node2)
        """
        self.goal_node = goal_node

        # min heap for the fringe
        self.fringe = PriorityQueue()
        self.fringe.add_node( root_node, self.h_score(root_node) )

        # min cost to node
        self.min_node_score = {}
        self.min_node_score[ str(root_node) ] = 0

        # path to start
        self.came_from = {}

        self.nodes_visited = 0

    def search(self):
        """
        Returns a list that contains the nodes of the shortest path to the goal,
        starting with the root node and ending with the goal node. Will return
        an empty list if there is no solution
        """
        while self.fringe:
            # grab the best candidate so far
            current_node = self.fringe.pop_node()

            self.nodes_visited += 1

            if self.is_goal(current_node): return self.path_to_root(current_node)
            else:
                self.add_children_to_fringe(current_node)
        # no path was found
        return []

    def h_score(self, node):
        """
        Returns the estimated cost from node to goal_node
        """
        raise NotImplementedError

    def children(self, node):
        """
        Returns a list of tuples with the first tuple value being the child
        and the second value being the cost to move to that child
        """
        raise NotImplementedError

    def path_to_root(self, node):
        """
        Return a list of nodes from root_node to node
        """
        path = [ node ]
        current_key = str(node)

        while current_key in self.came_from:
            next_node = self.came_from[current_key]
            path.append(next_node)
            current_key = str(next_node)

        # reverse order so root node is first
        path.reverse()

        for i, node in enumerate(path):
            if i == 0:
                print "Input  : %s" % node
            else:
                print "Move %2s: %s" % (i, node)

        return path


    def add_children_to_fringe(self, node):
        """
        Adds all the children of node to fringe and updates min_node_score and
        came_from dictionaries
        """
        node_key = str(node)
        cost_to_node = self.min_node_score[ node_key ]

        for child, cost_to_move in self.children(node):
            cost_to_child = cost_to_node + cost_to_move
            child_key = str(child)
            if child_key not in self.min_node_score or self.min_node_score[child_key] > cost_to_child:
                self.fringe.add_node(child, cost_to_child + self.h_score(child))
                self.min_node_score[child_key] = cost_to_child
                self.came_from[child_key] = node

    def is_goal(self, node):
        """
        Returns True if node is the same as goal_node, False otherwise
        """
        return str(node) == str(self.goal_node)

